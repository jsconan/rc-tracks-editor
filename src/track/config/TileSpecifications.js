/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Gives the specifications for the track tiles with respect to the initial constraints.
 */
export class TileSpecifications {
    /**
     * Defines the specifications for the track tiles from the given constraints.
     * @param {number} laneWidth - The width of the track lane (the distance between the barriers).
     * @param {number} barrierWidth - The width of the barriers.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @param {number} maxRatio - The maximum value for size ratios.
     */
    constructor(laneWidth = 20, barrierWidth = 1, barrierChunks = 4, maxRatio = 4) {
        this.setLaneWidth(laneWidth);
        this.setBarrierWidth(barrierWidth);
        this.setBarrierChunks(barrierChunks);
        this.setMaxRatio(maxRatio);
    }

    /**
     * The length of a tile with respect to the initial constraints.
     * @type {number}
     */
    get length() {
        return this.laneWidth * 1.25 + this.barrierWidth * 2;
    }

    /**
     * The width of a tile with respect to the initial constraints.
     * @type {number}
     */
    get width() {
        return this.laneWidth + this.barrierWidth * 2;
    }

    /**
     * The length of a barrier chunk with respect to the initial constraints.
     * @type {number}
     */
    get barrierLength() {
        return this.length / this.barrierChunks;
    }

    /**
     * The padding around the width of the tile to align with its length.
     * Note: tiles must fit in a square so that they can be flipped and rotated indifferently.
     * @type {number}
     */
    get padding() {
        return this.laneWidth / 8;
    }

    /**
     * Sets the width of the track lane (the distance between the barriers).
     * @param {number} laneWidth - The width of the track lane.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setLaneWidth(laneWidth) {
        this.laneWidth = Math.abs(laneWidth);

        return this;
    }

    /**
     * Sets the width of the barriers.
     * @param {number} barrierWidth - The width of the barriers.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierWidth(barrierWidth) {
        this.barrierWidth = Math.abs(barrierWidth);

        return this;
    }

    /**
     * Sets the number of barrier chunks per tile.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierChunks(barrierChunks) {
        this.barrierChunks = Math.abs(Math.round(barrierChunks) || 1);

        return this;
    }

    /**
     * Sets the maximum value for size ratios.
     * @param {number} maxRatio - The maximum value for size ratios.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setMaxRatio(maxRatio) {
        this.maxRatio = Math.abs(Math.round(maxRatio) || 1);

        return this;
    }
}
