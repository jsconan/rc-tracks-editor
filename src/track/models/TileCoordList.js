/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { writable } from 'svelte/store';
import { validateCallback } from '../../core/helpers';
import { TileList } from './TileList.js';

/**
 * Represents a list of tile coordinates.
 * It binds a list of tiles with a builder.
 */
export class TileCoordList {
    /**
     * Binds a list of tiles and a builder in order to produce a list of tile coordinates.
     * @param {TileList} list - The list of tiles to bind with the builder.
     * @param {function} builder - The reference to the builder.
     * @param {object} [options] - Some options for the builder.
     * @throws {TypeError} - If the given list is not a valid instance of TileList.
     * @throws {TypeError} - If the given builder is not a function.
     */
    constructor(list, builder, options = {}) {
        const { subscribe, set } = writable();
        let unsubscribe;

        /**
         * Adds a subscriber that will be notified each time the coordinates are modified.
         * @function subscribe
         * @param {function} subscriber - A callback that will receive notifications when the coordinates change.
         * @returns {function} - Returns a callback for removing the subscription.
         */
        this.subscribe = subscribe;

        /**
         * Computes the coordinates of each tile in the list.
         * Notifies all subscribers.
         * @returns {TileCoordList} - Chains the instance.
         */
        this.build = () => {
            if (this.list && this.builder) {
                this.listCoord = this.builder(this.list, this.options);
                set(this.listCoord);
            }

            return this;
        };

        /**
         * Binds the list of tiles with the builder.
         * @param {TileList} newList - The list of tiles to bind with the builder.
         * @returns {TileCoordList} - Chains the instance.
         * @throws {TypeError} - If the given list is not a valid instance of TileList.
         */
        this.setList = newList => {
            TileList.validateInstance(newList);

            if (unsubscribe) {
                unsubscribe();
            }

            this.list = newList;

            unsubscribe = newList.subscribe(this.build);

            return this;
        };

        this.setOptions(options);
        this.setList(list);
        this.setBuilder(builder);
    }

    /**
     * Binds the builder.
     * @param {function} builder - The reference to the builder.
     * @returns {TileCoordList} - Chains the instance.
     * @throws {TypeError} - If the given builder is not a function.
     */
    setBuilder(builder) {
        validateCallback(builder, 'builder');

        /**
         * Process a list of tiles for rendering, computing the coordinates of each tile.
         * @function builder
         * @param {TileList} list - The list of tiles from which compute the coordinates of each tile.
         * @param {object} [config] - A set of config options.
         * @returns {listCoord} - Returns the list of coordinates.
         */
        this.builder = builder;

        return this.build();
    }

    /**
     * Sets the options that will be given to the builder on the next call.
     * @param {object} options - Some options for the builder.
     * @returns {TileCoordList} - Chains the instance.
     */
    setOptions(options) {
        this.options = Object.assign({}, options);

        return this.build();
    }

    /**
     * Sets an option that will be given to the builder on the next call.
     * @param {string} name - The name of the option to set.
     * @param {*} value - The value for the option.
     * @returns {TileCoordList} - Chains the instance.
     */
    setOption(name, value) {
        this.options[name] = value;

        return this.build();
    }

    /**
     * Gets the value of a particular option for the builder.
     * @param {string} name - The name of the option to read.
     * @returns {*} - The value of the option.
     */
    getOption(name) {
        return this.options[name];
    }

    /**
     * Tells if an option for the builder has a value assigned.
     * @param {string} name - The name of the option to test.
     * @returns {boolean} - Returns `true` if the option has a value.
     */
    hasOption(name) {
        return Object.prototype.hasOwnProperty.call(this.options, name);
    }

    /**
     * Gets the position of a tile inside the list.
     * @param {string} id - The unique identifier of the tile.
     * @returns {number} - The position of the tile, or `-1` if it does not exist.
     */
    getTileIndex(id) {
        return this.list.getTileIndex(id);
    }

    /**
     * Retrieves the coordinates of a tile by its identifier.
     * @param {string} id - The unique identifier of the tile.
     * @returns {tileCoord} - The coordinates of the tile, or `null` if it does not exist.
     */
    getTile(id) {
        return this.getTileAt(this.getTileIndex(id));
    }

    /**
     * Gets the coordinates of a tile from a particular position in the list.
     * @param {number} index - The position of the tile.
     * @returns {tileCoord} - The coordinates of the tile, or `null` if it does not exist.
     */
    getTileAt(index) {
        if (!this.listCoord) {
            return null;
        }

        return this.listCoord.tiles[index] || null;
    }

    /**
     * Validates that the given model is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} model - The instance to validate.
     * @throws {TypeError} - If the given model is not a valid instance.
     */
    static validateInstance(model) {
        if (!(model instanceof this)) {
            throw new TypeError(`The model must be an instance of ${this.name}!`);
        }
    }
}

/**
 * @typedef {object} listCoord - Represents a list of tiles ready to be rendered.
 * @property {number} x - The left coordinate of the list of tiles.
 * @property {number} y - The top coordinate of the list of tiles.
 * @property {number} width - The width of the list of tiles.
 * @property {number} height - The height of the list of tiles.
 * @property {tileCoord[]} tiles - The list of tiles.
 * @property {Counter} [stats] - An object listing the stats for the list of tiles.
 */

/**
 * @typedef {object} tileCoord - Represents a positioned tile.
 * @property {string} id - The unique identifier of the tile.
 * @property {string} type - The type of tile.
 * @property {string} direction - The direction of the tile.
 * @property {number} ratio - The size ratio of the tile.
 * @property {number} x - The left coordinate of the tile.
 * @property {number} y - The top coordinate of the tile.
 * @property {number} angle - The rotation angle of the tile.
 * @property {tileRect} rect - The bounding rectangle of the tile.
 * @property {TileModel} model - A reference to the tile model.
 */

/**
 * @typedef {import('../models/TileModel.js').tileRect} tileRect
 */

/**
 * @typedef {import('../models').TileModel} TileModel
 */

/**
 * @typedef {import('../../core/models').Counter} Counter
 */
