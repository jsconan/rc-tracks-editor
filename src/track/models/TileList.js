/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {
    validateType,
    CURVED_TILE_ENLARGED_TYPE,
    CURVED_TILE_TYPE,
    STRAIGHT_TILE_TYPE,
    TILE_DIRECTION_RIGHT
} from '../helpers';
import { uid } from '../../core/helpers';
import { CurvedTileEnlargedModel } from './CurvedTileEnlargedModel.js';
import { CurvedTileModel } from './CurvedTileModel.js';
import { StraightTileModel } from './StraightTileModel.js';
import { TileSpecifications } from '../config';
import { List } from '../../core/models';
import { eventStore } from '../../core/stores';

/**
 * @type {object} - Maps the types of tile to their respective model.
 * @private
 */
const modelsMap = {
    [STRAIGHT_TILE_TYPE]: StraightTileModel,
    [CURVED_TILE_TYPE]: CurvedTileModel,
    [CURVED_TILE_ENLARGED_TYPE]: CurvedTileEnlargedModel
};

/**
 * Creates the tile model with respect to the given type.
 * @param {TileSpecifications} specs - The specifications for the tiles.
 * @param {string} type - The type of tile.
 * @param {string} direction - The direction of the tile.
 * @param {number} ratio - The size ratio of the tile.
 * @returns {TileModel} - Returns a tile model of the expected type.
 * @throws {TypeError} - If the given specifications object is not valid.
 * @throws {TypeError} - If the given type is not valid.
 * @throws {TypeError} - If the given direction is not valid.
 * @private
 */
function createModel(specs, type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
    validateType(type);
    const model = new modelsMap[type](specs, direction, ratio);
    model.id = uid();
    return model;
}

/**
 * Represents a list of tiles.
 */
export class TileList {
    /**
     * Represents a list of tiles with the given size constraints.
     * @param {TileSpecifications} specs - The specifications for the tiles.
     * @throws {TypeError} - If the given specifications object is not valid.
     */
    constructor(specs) {
        this.tiles = new List();

        this.setSpecs(specs);

        // Produces a store linked to the list events and returning the TileList
        const { subscribe } = eventStore(
            this.tiles,
            [
                // List events
                'set',
                'insert',
                'add',
                'delete',
                'clear',
                'load',
                // TileList events
                'specs',
                'update'
            ],
            () => this
        );

        /**
         * Adds a subscriber that will be notified each time the list is modified.
         * @function subscribe
         * @param {function} subscriber - A callback that will receive notifications when the list is changed.
         * @returns {function} - Return a callback for removing the subscription.
         */
        this.subscribe = subscribe;
    }

    /**
     * The number of tiles in the list.
     * @type {number}
     */
    get length() {
        return this.tiles.length;
    }

    /**
     * Iterates over the tiles from the list.
     * @yields {TileModel} - The next tile in the list.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.tiles.values();
    }

    /**
     * Returns an iterator for the tiles from the list.
     * @yields {TileModel} - The next tile in the list.
     * @generator
     */
    *values() {
        yield* this.tiles.values();
    }

    /**
     * Sets the specifications for the tiles.
     * @param {TileSpecifications} specs - The specifications for the tiles.
     * @returns {TileList} - Chains the instance.
     * @throws {TypeError} - If the given specifications object is not valid.
     * @fires specs
     */
    setSpecs(specs) {
        TileSpecifications.validateInstance(specs);

        this.specs = specs;

        if (this.tiles.length) {
            this.tiles.forEach(tile => tile.setSpecs(specs));

            /**
             * Notifies the tile specification have changed.
             * @event specs
             * @param {TileSpecifications} specs - The specifications for the tiles.
             */
            this.tiles.emit('specs', specs);
        }

        return this;
    }

    /**
     * Notifies an update of the tiles.
     * This is useful when one or several tiles have been modified internally without calling the list API.
     * @returns {TileList} - Chains the instance.
     * @fires update
     */
    update() {
        /**
         * Notifies a change in the tiles.
         * @event update
         */
        this.tiles.emit('update');

        return this;
    }

    /**
     * Gets the position of a tile inside the list.
     * @param {string} id - The unique identifier of the tile.
     * @returns {number} - The position of the tile, or `-1` if it does not exist.
     */
    getTileIndex(id) {
        return this.tiles.find(tile => tile.id === id);
    }

    /**
     * Retrieves a tile by its identifier.
     * @param {string} id - The unique identifier of the tile.
     * @returns {TileModel} - The referenced tile, or `null` if it does not exist.
     */
    getTile(id) {
        return this.getTileAt(this.getTileIndex(id));
    }

    /**
     * Gets a tile from a particular position.
     * @param {number} index - The position of the tile.
     * @returns {TileModel} - The referenced tile, or `null` if it does not exist.
     */
    getTileAt(index) {
        return this.tiles.get(index) || null;
    }

    /**
     * Add a tile to the list, at the last position.
     * @param {string} type - The type of tile to add.
     * @param {string} direction - The direction of the tile, can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @returns {string} - Returns the unique identifier of the added tile.
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     * @fires add
     */
    appendTile(type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        const tile = createModel(this.specs, type, direction, ratio);

        this.tiles.add(tile);

        return tile.id;
    }

    /**
     * Add a tile to the list, at the first position.
     * @param {string} type - The type of tile to add.
     * @param {string} direction - The direction of the tile, can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @returns {string} - Returns the unique identifier of the added tile.
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     * @fires insert
     */
    prependTile(type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        const tile = createModel(this.specs, type, direction, ratio);

        this.tiles.insert(0, tile);

        return tile.id;
    }

    /**
     * Removes a tile from the list.
     * @param {string} id - The unique identifier of the tile to remove.
     * @returns {boolean} - Returns `true` if the deletion succeeds. Otherwise, returns `false`.
     * @fires delete
     */
    removeTile(id) {
        const index = this.getTileIndex(id);

        if (index < 0) {
            return false;
        }

        return this.tiles.delete(index) > 0;
    }

    /**
     * Replace a tile in the list.
     * If the tile does not exist, it does nothing.
     * @param {string} id - The unique identifier of the tile to replace.
     * @param {string} type - The type of tile to add.
     * @param {string} direction - The direction of the tile, can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @returns {string} - Returns the unique identifier of the added tile.
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     * @fires set
     */
    replaceTile(id, type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        const index = this.getTileIndex(id);

        if (index >= 0) {
            const tile = createModel(this.specs, type, direction, ratio);

            this.tiles.set(index, tile);

            return tile.id;
        }

        return null;
    }

    /**
     * Insert a tile in the list before a particular position.
     * If the position does not exist, it does nothing.
     * @param {string} id - The unique identifier of the tile before which add another tile.
     * @param {string} type - The type of tile to add.
     * @param {string} direction - The direction of the tile, can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @returns {string} - Returns the unique identifier of the added tile.
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     * @fires insert
     */
    insertTileBefore(id, type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        const index = this.getTileIndex(id);

        if (index >= 0) {
            const tile = createModel(this.specs, type, direction, ratio);

            this.tiles.insert(index, tile);

            return tile.id;
        }

        return null;
    }

    /**
     * Insert a tile in the list after a particular position.
     * If the position does not exist, it does nothing.
     * @param {string} id - The unique identifier of the tile after which add another tile.
     * @param {string} type - The type of tile to add.
     * @param {string} direction - The direction of the tile, can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @returns {string} - Returns the unique identifier of the added tile.
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     * @fires insert
     */
    insertTileAfter(id, type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        const index = this.getTileIndex(id);

        if (index >= 0) {
            const tile = createModel(this.specs, type, direction, ratio);

            this.tiles.insert(index + 1, tile);

            return tile.id;
        }

        return null;
    }

    /**
     * Exports the list to an object.
     * @returns {tileExport[]} - An object representation of the list.
     */
    export() {
        return this.tiles.map(tile => tile.export());
    }

    /**
     * Imports the list from a source.
     * The list is emptied before importing, any existing tile will be deleted.
     * @param {tileExport[]} data - An iterable object containing a representation of the list.
     * @returns {TileList} - Chains the instance.
     * @fires load
     */
    import(data) {
        if (!data || !data[Symbol.iterator]) {
            return this;
        }

        const dataIterator = data[Symbol.iterator]();
        const loadIterator = {
            next: () => {
                const next = dataIterator.next();

                if (next.done) {
                    return next;
                }

                const { type, direction, ratio } = next.value || {};
                const tile = createModel(this.specs, type, direction, ratio);

                return { done: false, value: tile };
            },

            [Symbol.iterator]() {
                return this;
            }
        };

        this.tiles.load(loadIterator);

        return this;
    }

    /**
     * Removes all tiles from the list.
     * @returns {TileList} - Chains the instance.
     * @fires clear
     */
    clear() {
        this.tiles.clear();

        return this;
    }

    /**
     * Validates that the given model is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} model - The instance to validate.
     * @throws {TypeError} - If the given model is not a valid instance.
     */
    static validateInstance(model) {
        if (!(model instanceof this)) {
            throw new TypeError(`The model must be an instance of ${this.name}!`);
        }
    }
}

/**
 * @typedef {import('./TileModel').tileExport} tileExport
 */

/**
 * @typedef {import('./TileModel').TileModel} TileModel
 */
