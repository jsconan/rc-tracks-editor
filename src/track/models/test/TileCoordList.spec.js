/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { TileSpecifications } from '../../config';
import { TileList } from '../TileList.js';
import { TileCoordList } from '../TileCoordList.js';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications(laneWidth, barrierWidth, barrierChunks);

const mockBuilder = list => {
    const tiles = list.tiles.map(tile => `coord-${tile.id}`);
    return { tiles };
};

describe('TileCoordList', () => {
    it('is a class', () => {
        expect(TileCoordList).toEqual(expect.any(Function));
    });

    it('has a length', () => {
        const list = new TileList(specs);
        const listCoord = new TileCoordList(list, mockBuilder);

        expect(listCoord.length).toBe(0);

        list.appendTile();

        expect(listCoord.length).toBe(1);
    });

    it('implements the iteration protocol', () => {
        const list = new TileList(specs);
        const listCoord = new TileCoordList(list, mockBuilder);
        list.appendTile();

        expect(listCoord[Symbol.iterator]).toEqual(expect.any(Function));
        expect(listCoord[Symbol.iterator]()).not.toBe(listCoord[Symbol.iterator]());
        expect([...listCoord]).toMatchSnapshot();
    });

    it('can produce an iterator', () => {
        const list = new TileList(specs);
        const listCoord = new TileCoordList(list, mockBuilder);
        list.appendTile();

        expect(listCoord.values).toEqual(expect.any(Function));
        expect(listCoord.values()[Symbol.iterator]).toEqual(expect.any(Function));

        expect(listCoord.values()).not.toBe(listCoord.values());
        expect([...listCoord.values()]).toMatchSnapshot();
    });

    describe('throws error', () => {
        it('when trying to create an instance with an invalid list', () => {
            // @ts-expect-error
            expect(() => new TileCoordList({})).toThrow('The model must be an instance of TileList!');
        });

        it('when trying to create an instance with an invalid builder', () => {
            // @ts-expect-error
            expect(() => new TileCoordList(new TileList(specs), {})).toThrow('A builder function is expected!');
        });

        it('when trying to set an invalid list', () => {
            const listCoord = new TileCoordList(new TileList(specs), mockBuilder);
            // @ts-expect-error
            expect(() => listCoord.setList({})).toThrow('The model must be an instance of TileList!');
        });

        it('when trying to set an invalid builder', () => {
            const listCoord = new TileCoordList(new TileList(specs), mockBuilder);
            // @ts-expect-error
            expect(() => listCoord.setBuilder({})).toThrow('A builder function is expected!');
        });
    });

    describe('can set', () => {
        it('a list of tiles', () => {
            const oldList = new TileList(specs);
            const newList = new TileList(specs);
            const listCoord = new TileCoordList(oldList, mockBuilder);

            expect(listCoord.list).toBe(oldList);
            expect(listCoord.setList(newList)).toBe(listCoord);
            expect(listCoord.list).toBe(newList);
        });

        it('a builder function', () => {
            const listCoord = new TileCoordList(new TileList(specs), mockBuilder);
            const builder = () => {};

            expect(listCoord.builder).not.toBe(builder);
            expect(listCoord.setBuilder(builder)).toBe(listCoord);
            expect(listCoord.builder).toBe(builder);
        });

        it('the options for the builder', () => {
            const listCoord = new TileCoordList(new TileList(specs), mockBuilder, { dummy: true });
            const options = { foo: 'bar' };

            expect(listCoord.hasOption('dummy')).toBeTruthy();
            expect(listCoord.setOptions(options)).toBe(listCoord);
            expect(listCoord.options).not.toBe(options);
            expect(listCoord.hasOption('dummy')).toBeFalsy();
            expect(listCoord.getOption('foo')).toBe(options.foo);
        });
    });

    describe('manage options for the builder', () => {
        it('setting the value of an option', () => {
            const listCoord = new TileCoordList(new TileList(specs), mockBuilder);

            expect(listCoord.hasOption('foo')).toBeFalsy();
            expect(listCoord.setOption('foo', 'bar')).toBe(listCoord);
            expect(listCoord.hasOption('foo')).toBeTruthy();
            expect(listCoord.getOption('foo')).toBe('bar');
        });

        it('getting the value of an option', () => {
            const listCoord = new TileCoordList(new TileList(specs), mockBuilder);

            expect(listCoord.getOption('foo')).toBeUndefined();
            listCoord.setOption('foo', 'bar');
            expect(listCoord.getOption('foo')).toBe('bar');
        });

        it('checking if an option is assigned', () => {
            const listCoord = new TileCoordList(new TileList(specs), mockBuilder);

            expect(listCoord.hasOption('foo')).toBeFalsy();
            listCoord.setOption('foo', 'bar');
            expect(listCoord.hasOption('foo')).toBeTruthy();
        });
    });

    describe('manage the built coordinates', () => {
        const list = new TileList(specs);
        const listCoord = new TileCoordList(list, mockBuilder);

        list.appendTile();
        const id = list.appendTile();
        list.appendTile();

        it('getting the identifier of a tile at a particular index', () => {
            expect(listCoord.getTileIndex(id)).toBe(1);
            expect(listCoord.getTileIndex('foo')).toBe(-1);
        });

        it('getting the coordinate of a tile from its identifier', () => {
            expect(listCoord.getTile(id)).toBe(`coord-${id}`);
            expect(listCoord.getTile('foo')).toBeNull();
        });

        it('getting the coordinate of a tile at a particular index', () => {
            expect(listCoord.getTileAt(1)).toBe(`coord-${id}`);
            expect(listCoord.getTileAt(3)).toBeNull();

            const emptyBridge = new TileCoordList(new TileList(specs), () => {});
            expect(emptyBridge.getTileAt(0)).toBeNull();
        });
    });

    describe('can build coordinates', () => {
        it('with the given builder options', () => {
            const theList = new TileList(specs);
            theList.appendTile();

            const builder = (list, options) => {
                expect(list).toBe(theList);
                expect(options).toEqual(expect.any(Object));
                expect(options.foo).toBe('bar');
                return mockBuilder(list);
            };

            const listCoord = new TileCoordList(theList, builder, { foo: 'bar' });

            const callback = jest.fn().mockImplementation(() => {
                expect(listCoord.listCoord).toMatchSnapshot();
            });
            const unsubscribe = listCoord.subscribe(callback);

            expect(callback).toHaveBeenCalledTimes(1);
            unsubscribe();
        });

        it('when the list of tiles is changed', () => {
            const oldList = new TileList(specs);
            oldList.appendTile();

            const newList = new TileList(specs);
            newList.appendTile();

            const listCoord = new TileCoordList(oldList, mockBuilder);

            const callback = jest.fn().mockImplementation(() => {
                expect(listCoord.listCoord).toMatchSnapshot();
            });
            const unsubscribe = listCoord.subscribe(callback);

            expect(listCoord.setList(newList)).toBe(listCoord);
            oldList.appendTile();
            newList.appendTile();

            expect(callback).toHaveBeenCalledTimes(3);

            unsubscribe();
        });

        it('when the builder function is changed', () => {
            const list = new TileList(specs);
            list.appendTile();

            const listCoord = new TileCoordList(list, mockBuilder);

            const callback = jest.fn().mockImplementation(() => {
                expect(listCoord.listCoord).toMatchSnapshot();
            });
            const unsubscribe = listCoord.subscribe(callback);

            const newBuilder = () => list.tiles.map(tile => `coord-${tile.id}`);
            expect(listCoord.setBuilder(newBuilder)).toBe(listCoord);

            expect(callback).toHaveBeenCalledTimes(2);

            unsubscribe();
        });

        it('each time a tile is added', () => {
            const theList = new TileList(specs);
            const listCoord = new TileCoordList(theList, mockBuilder);

            const callback = jest.fn().mockImplementation(() => {
                expect(listCoord.listCoord).toMatchSnapshot();
            });
            const unsubscribe = listCoord.subscribe(callback);

            theList.appendTile();

            expect(callback).toHaveBeenCalledTimes(2);
            unsubscribe();
        });

        it('each time an option is changed', () => {
            const theList = new TileList(specs);
            theList.appendTile();

            const listCoord = new TileCoordList(theList, mockBuilder);

            const callback = jest.fn().mockImplementation(() => {
                expect(listCoord.listCoord).toMatchSnapshot();
            });
            const unsubscribe = listCoord.subscribe(callback);

            listCoord.setOption('foo', 'bar');
            listCoord.setOptions({});

            expect(callback).toHaveBeenCalledTimes(3);
            unsubscribe();
        });

        it('each time the build function is called', () => {
            const theList = new TileList(specs);
            theList.appendTile();

            const listCoord = new TileCoordList(theList, mockBuilder);

            const callback = jest.fn().mockImplementation(() => {
                expect(listCoord.listCoord).toMatchSnapshot();
            });
            const unsubscribe = listCoord.subscribe(callback);

            listCoord.build();

            expect(callback).toHaveBeenCalledTimes(2);
            unsubscribe();
        });
    });

    it('can validate an object is an instance of the class', () => {
        const list = new TileList(specs);
        const listCoord = new TileCoordList(list, mockBuilder);
        expect(() => TileCoordList.validateInstance(listCoord)).not.toThrow();
        expect(() => TileCoordList.validateInstance({})).toThrow('The model must be an instance of TileCoordList!');
    });
});
