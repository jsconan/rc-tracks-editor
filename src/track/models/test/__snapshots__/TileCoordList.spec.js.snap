// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileCoordList can build coordinates each time a tile is added 1`] = `
Object {
  "tiles": Array [],
}
`;

exports[`TileCoordList can build coordinates each time a tile is added 2`] = `
Object {
  "tiles": Array [
    "coord-id-12",
  ],
}
`;

exports[`TileCoordList can build coordinates each time an option is changed 1`] = `
Object {
  "tiles": Array [
    "coord-id-13",
  ],
}
`;

exports[`TileCoordList can build coordinates each time an option is changed 2`] = `
Object {
  "tiles": Array [
    "coord-id-13",
  ],
}
`;

exports[`TileCoordList can build coordinates each time an option is changed 3`] = `
Object {
  "tiles": Array [
    "coord-id-13",
  ],
}
`;

exports[`TileCoordList can build coordinates each time the build function is called 1`] = `
Object {
  "tiles": Array [
    "coord-id-14",
  ],
}
`;

exports[`TileCoordList can build coordinates each time the build function is called 2`] = `
Object {
  "tiles": Array [
    "coord-id-14",
  ],
}
`;

exports[`TileCoordList can build coordinates when the builder function is changed 1`] = `
Object {
  "tiles": Array [
    "coord-id-11",
  ],
}
`;

exports[`TileCoordList can build coordinates when the builder function is changed 2`] = `
Array [
  "coord-id-11",
]
`;

exports[`TileCoordList can build coordinates when the list of tiles is changed 1`] = `
Object {
  "tiles": Array [
    "coord-id-7",
  ],
}
`;

exports[`TileCoordList can build coordinates when the list of tiles is changed 2`] = `
Object {
  "tiles": Array [
    "coord-id-8",
  ],
}
`;

exports[`TileCoordList can build coordinates when the list of tiles is changed 3`] = `
Object {
  "tiles": Array [
    "coord-id-8",
    "coord-id-10",
  ],
}
`;

exports[`TileCoordList can build coordinates with the given builder options 1`] = `
Object {
  "tiles": Array [
    "coord-id-6",
  ],
}
`;

exports[`TileCoordList can produce an iterator 1`] = `
Array [
  "coord-id-5",
]
`;

exports[`TileCoordList implements the iteration protocol 1`] = `
Array [
  "coord-id-4",
]
`;
